### PR TITLE
Source GitHub: support self hosted instances

### LINKED ISSUE #24007: `source-github` should support self-hosted Github instances
Right now, `source-github` only connects to `api.github.com`.  To support self-hosted instances, we should:
* add an optional option to the connectors SPEC, `api_url ` which has a default value of `api.github.com` (to make this change backwards compatible).
* Update the code to use the new config option
* Add `"${api_url}"` to the connector's allowed_hosts

### PR DESCRIPTION
**Pull request by artem1205**
## What

Resolve https://github.com/airbytehq/airbyte/issues/24007

## How

add support for self-hosted instances

## Recommended reading order
1. `source.py`

## 🚨 User Impact 🚨

no breaking changes

## Pre-merge Actions
<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

### CODE REVIEW

**Comment by airbyte-oss-build-runner** [comment]
##  source-github test report (commit [`99016b353a`](https://github.com/airbytehq/airbyte/pull/30647/commits/99016b353a4de44456e987cc0d716b1e3fede588)) - ❌

⏲️  Total pipeline duration: 06mn04s 

| Step                                                                 | Result   |
|:---------------------------------------------------------------------|:---------|
| Connector package install                                            | ✅        |
| Build source-github docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                           | ✅        |
| Acceptance tests                                                     | ❌        |
| Code format checks                                                   | ✅        |
| Validate airbyte-integrations/connectors/source-github/metadata.yaml | ✅        |
| Connector version semver check                                       | ✅        |
| Connector version increment check                                    | ✅        |
| QA checks                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/artem1205_source-github-support-self-hosted/1695384883/99016b353a4de44456e987cc0d716b1e3fede588/source-github/1.2.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:99016b353a4de44456e987cc0d716b1e3fede588)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-github test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-github test report (commit [`eaf7e4ad6e`](https://github.com/airbytehq/airbyte/pull/30647/commits/eaf7e4ad6ece7a490c97c2d3d2671e470829e32f)) - ✅

⏲️  Total pipeline duration: 05mn59s 

| Step                                                                 | Result   |
|:---------------------------------------------------------------------|:---------|
| Connector package install                                            | ✅        |
| Build source-github docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                           | ✅        |
| Acceptance tests                                                     | ✅        |
| Code format checks                                                   | ✅        |
| Validate airbyte-integrations/connectors/source-github/metadata.yaml | ✅        |
| Connector version semver check                                       | ✅        |
| Connector version increment check                                    | ✅        |
| QA checks                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/artem1205_source-github-support-self-hosted/1695386587/eaf7e4ad6ece7a490c97c2d3d2671e470829e32f/source-github/1.2.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:eaf7e4ad6ece7a490c97c2d3d2671e470829e32f)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-github test
```

File: airbyte-integrations/connectors/source-github/source_github/source.py [hunk 1]
```diff
@@ -139,6 +142,26 @@ def _get_authenticator(self, config: Mapping[str, Any]):
...
+            message = "Please enter full url starting from http..."
+        elif api_url_parsed.scheme == "http" and not self._is_http_allowed():
+            message = "HTTP connection is insecure an is not allowed in this environment. Please use `https` instead."
+        elif not api_url_parsed.netloc:
+            message = "Please provide correct URL"
+        else:
+            message = None
+        if message:
+            raise AirbyteTracedException(message=message, failure_type=FailureType.config_error)
+        return config
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-github/source_github/source.py:159]:
```suggestion
        config.setdefault("api_url", "https://api.github.com")
        api_url_parsed = urlparse(config["api_url"])

        if not api_url_parsed.scheme.startswith("http"):
            message = "Please enter a full URL starting with http..."
        elif api_url_parsed.scheme == "http" and not self._is_http_allowed():
            message = "HTTP connection is insecure and is not allowed in this environment. Please use `https` instead."
        elif not api_url_parsed.netloc:
            message = "Please provide a correct URL"
        else:
            return config

        raise AirbyteTracedException(message=message, failure_type=FailureType.config_error)
```

**Review by lazebnyi — APPROVED** [review]
LGTM, need small refactoring for `_ensure_default_values` method.

**Comment by airbyte-oss-build-runner** [comment]
##  source-github test report (commit [`2bd4b5d5c2`](https://github.com/airbytehq/airbyte/pull/30647/commits/2bd4b5d5c24684d7f23cfe425ead9285bff1a658)) - ❌

⏲️  Total pipeline duration: 01mn01s 

| Step                                                                 | Result   |
|:---------------------------------------------------------------------|:---------|
| Connector package install                                            | ✅        |
| Build source-github docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                           | ❌        |
| Code format checks                                                   | ✅        |
| Validate airbyte-integrations/connectors/source-github/metadata.yaml | ✅        |
| Connector version semver check                                       | ✅        |
| Connector version increment check                                    | ✅        |
| QA checks                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/artem1205_source-github-support-self-hosted/1695401698/2bd4b5d5c24684d7f23cfe425ead9285bff1a658/source-github/1.2.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:2bd4b5d5c24684d7f23cfe425ead9285bff1a658)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-github test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-github test report (commit [`af83d53396`](https://github.com/airbytehq/airbyte/pull/30647/commits/af83d53396d3b01d74c09bdeeb2f84b86a108df3)) - ✅

⏲️  Total pipeline duration: 04mn39s 

| Step                                                                 | Result   |
|:---------------------------------------------------------------------|:---------|
| Connector package install                                            | ✅        |
| Build source-github docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                           | ✅        |
| Acceptance tests                                                     | ✅        |
| Code format checks                                                   | ✅        |
| Validate airbyte-integrations/connectors/source-github/metadata.yaml | ✅        |
| Connector version semver check                                       | ✅        |
| Connector version increment check                                    | ✅        |
| QA checks                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/artem1205_source-github-support-self-hosted/1695403522/af83d53396d3b01d74c09bdeeb2f84b86a108df3/source-github/1.2.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:af83d53396d3b01d74c09bdeeb2f84b86a108df3)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-github test
```

**Comment by davydov-d** [comment]
> ## Suspect Issues
> This pull request was deployed and Sentry observed the following issues:
> 
> * ‼️ **requests.exceptions.JSONDecodeError: Expecting value: line 1 column 1 (char 0)** `/airbyte/integration_code/source_github/streams...` [View Issue](https://airbytehq.sentry.io/issues/4509301795/?referrer=github-pr-bot)
> 
> Did you find this useful? React with a 👍 or 👎

@artem1205 can you please take a look at this?

**Comment by artem1205** [comment]
@davydov-d 
https://github.com/airbytehq/oncall/issues/3083 is not related to changes introduced in this PR
working on it

**Comment by davydov-d** [comment]
Thank you!

**Pull request merged**